File: go/y2023/d03/solution.go
package main

import (
	"adventofcode/h"
	"fmt"
	"io"
	"log"
	"os"
	"time"
)

type solution struct {
	input                  h.StringGrid
	ans1, ans2, nrow, ncol int
	gearsM                 map[h.Pt][]int
}

func buildSolution(r io.Reader) *solution {
	lines, _ := h.LinesFromReader(r)
	return &solution{
		input:  h.StringGrid{Array: lines},
		nrow:   len(lines),
		ncol:   len(lines[0]),
		ans1:   0,
		ans2:   0,
		gearsM: make(map[h.Pt][]int),
	}
}

func isSymb(b byte) bool {
	if h.IsdigitBool(b) {
		return false
	}
	return b != '.'
}

func (s *solution) check_nearby(p h.Pt) int {
	for _, dir := range h.Dir8 {
		nxp := p.PMove(dir)
		if !s.input.IsInside(nxp) {
			continue
		}
		b := s.input.GetByte(nxp)
		if isSymb(b) {
			return 1
		}
	}
	return 0
}
func (s *solution) check_gear(p h.Pt) (gears []h.Pt) {
	for _, dir := range h.Dir8 {
		nxp := p.PMove(dir)
		if !s.input.IsInside(nxp) {
			continue
		}
		b := s.input.GetByte(nxp)
		if b == '*' {
			gears = append(gears, nxp)
		}
	}
	return
}

func (s *solution) traves() {
	for i := 0; i < s.nrow; i++ {
		for j := 0; j < s.ncol; j++ {

			pt := h.Pt{R: i, C: j}
			b := s.input.GetByte(pt)

			if h.IsdigitBool(b) {
				// start scanning
				num, valid := 0, 0

				// build num
				for h.IsdigitBool(b) {
					cur_d, _ := h.Isdigit(b)
					num = num*10 + cur_d
					valid += s.check_nearby(pt)
					j++
					pt = h.Pt{R: i, C: j}
					if !s.input.IsInside(pt) {
						break
					}
					b = s.input.GetByte(pt)
				}
				if valid > 0 {
					s.ans1 += num
				}
			}
		}
	}
}
func (s *solution) run1() {
	s.traves()
}

func (s *solution) traves2() {
	for i := 0; i < s.nrow; i++ {
		for j := 0; j < s.ncol; j++ {

			pt := h.Pt{R: i, C: j}
			b := s.input.GetByte(pt)

			if h.IsdigitBool(b) {
				// start scanning
				var (
					num   int
					gears = map[h.Pt]bool{}
				)

				// build num
				for h.IsdigitBool(b) {
					cur_d, _ := h.Isdigit(b)
					num = num*10 + cur_d
					for _, gear := range s.check_gear(pt) {
						gears[gear] = true
					}
					j++
					pt = h.Pt{R: i, C: j}
					if !s.input.IsInside(pt) {
						break
					}
					b = s.input.GetByte(pt)
				}
				for pt := range gears {
					s.gearsM[pt] = append(s.gearsM[pt], num)
				}
			}
		}
	}
}
func (s *solution) run2() {
	s.traves2()
	for _, ints := range s.gearsM {
		if len(ints) == 2 {
			s.ans2 += ints[0] * ints[1]
		}
	}
}

func (s *solution) res1() int {
	return s.ans1
}

func (s *solution) res2() int {
	return s.ans2
}

func part1(r io.Reader) int {
	s := buildSolution(r)
	s.run1()
	return s.res1()
}

func part2(r io.Reader) int {
	s := buildSolution(r)
	s.run2()
	return s.res2()
}

func main() {
	Input, err := os.Open("input.txt")
	if err != nil {
		log.Fatalf("fail open input.txt %v", err)
	}
	start := time.Now()
	result := part1(Input)
	elapsed := time.Since(start)
	fmt.Printf("p1 res 🙆-> %d (Time taken: %s)\n", result, elapsed)
	start = time.Now()
	result = part2(Input)
	elapsed = time.Since(start)
	fmt.Printf("p2 res 🙆-> %d (Time taken: %s)\n", result, elapsed)
}

File: python/y2023/d03/solution.py
import time
from pathlib import Path

import python.h as h


def is_symbol(ch: str) -> bool:
    return ch != "." and not ch.isdigit()


def part1(text: str) -> int:
    grid = h.parse(text)
    nrow, ncol = len(grid), len(grid[0])
    total = 0

    for i in range(nrow):
        j = 0
        while j < ncol:
            if grid[i][j].isdigit():
                num, valid = 0, False
                while j < ncol and grid[i][j].isdigit():
                    num = num * 10 + int(grid[i][j])
                    for dr, dc in h.DIR8:
                        nr, nc = i + dr, j + dc
                        if (
                            0 <= nr < nrow
                            and 0 <= nc < ncol
                            and is_symbol(grid[nr][nc])
                        ):
                            valid = True
                    j += 1
                if valid:
                    total += num
            else:
                j += 1
    return total


def part2(text: str) -> int:
    grid = h.parse(text)
    nrow, ncol = len(grid), len(grid[0])
    gears: dict[tuple[int, int], list[int]] = {}

    for i in range(nrow):
        j = 0
        while j < ncol:
            if grid[i][j].isdigit():
                num = 0
                adj_gears: set[tuple[int, int]] = set()
                while j < ncol and grid[i][j].isdigit():
                    num = num * 10 + int(grid[i][j])
                    for dr, dc in h.DIR8:
                        nr, nc = i + dr, j + dc
                        if 0 <= nr < nrow and 0 <= nc < ncol and grid[nr][nc] == "*":
                            adj_gears.add((nr, nc))
                    j += 1
                for g in adj_gears:
                    gears.setdefault(g, []).append(num)
            else:
                j += 1

    return sum(nums[0] * nums[1] for nums in gears.values() if len(nums) == 2)


def main():
    input_path = Path(__file__).parent / "input.txt"
    text = input_path.read_text()

    start = time.perf_counter()
    result = part1(text)
    elapsed = time.perf_counter() - start
    print(f"p1 res -> {result} (Time taken: {elapsed:.4f}s)")

    start = time.perf_counter()
    result = part2(text)
    elapsed = time.perf_counter() - start
    print(f"p2 res -> {result} (Time taken: {elapsed:.4f}s)")


if __name__ == "__main__":
    main()
